### PR TITLE
Add Card, Tag, Learner services

### PR DESCRIPTION
**Pull request by Chibbluffy**
Includes login page changes from that branch.
Includes 3 added services
Also filled out Tag model

### CODE REVIEW

**Comment by Chibbluffy** [comment]
Oh, forgot to mention the added dependencies (HttpClient, etc)

File: src/app/services/card/card.service.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  constructor(private http:HttpClient) { }
+
+  async createCard(card:Card){
+    card = await this.http.post<Card>(`http://34.122.220.146:8080/cards`,card).toPromise();
+    return card;
+  }
+
+  async getCardById(id:number){
+    const cards:Card[] = await this.http.get<Card[]>(`http://34.122.220.146:8080/cards/${id}`).toPromise();
+    return cards;
```
Comment (Lcarrico) [inline src/app/services/card/card.service.ts]: This should return a single card, not an array of cards.

File: src/app/models/tag.ts [hunk 2]
```diff
@@ -0,0 +1,11 @@
+import { Card } from "src/app/models/card";
+import { Stack } from "src/app/models/stack";
+
+export class Tag{
+    constructor(
+        public tagId:number,
+        public tagName:string,
+        public stacks:Stack[],
+        public cards:Card[]
```
Comment (Lcarrico) [inline src/app/models/tag.ts]: Please remove stacks and cards, the relationship will only be accessible through the stacks and cards and not the tags.

File: src/app/services/learner/learner.service.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+export class LearnerService {
+
+  constructor(private http:HttpClient) { }
+
+  async createLearner(learner:Learner){
+    learner = await this.http.post<Learner>(`http://34.122.220.146:8080/learners`,learner).toPromise();
+    return learner;
+  }
+
+  async getLearnerByUsername(id:number){
```
Comment (Lcarrico) [inline src/app/services/learner/learner.service.ts]: This should be getLearnerById not getLearnerByUsername.

File: src/app/services/learner/learner.service.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    return learners;
+  }
+
+  async updateLearner(learner:Learner){
+    learner = await this.http.put<Learner>(`http://34.122.220.146:8080/learners`,learner).toPromise();
+    return learner;
+  }
+
+  async deleteCard(learner:Learner){
+    return await this.http.delete(`http://34.122.220.146:8080/learners/${learner.cardId}`).toPromise();
```
Comment (Lcarrico) [inline src/app/services/learner/learner.service.ts]: Please re-write this function.

Comment (Lcarrico) [reply src/app/services/learner/learner.service.ts]: Name is wrong and the request URI is wrong as well.

File: src/app/services/tag/tag.service.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+  }
+
+  async getTagById(id:number){
+    const tags:Tag[] = await this.http.get<Tag[]>(`http://34.122.220.146:8080/tags/${id}`).toPromise();
+    return tags;
+  }
+
+  async getTagByName(tagName:string){
+    const tags:Tag[] = await this.http.get<Tag[]>(`http://34.122.220.146:8080/tags?tagName=${tagName}`).toPromise();
+    return tags;
```
Comment (Lcarrico) [inline src/app/services/tag/tag.service.ts:26]: This should return the first element of the tags array.

File: src/app/services/tag/tag.service.ts [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+export class TagService {
+
+  constructor(private http:HttpClient) { }
+
+  async createTag(tag:Tag){
+    tag = await this.http.post<Tag>(`http://34.122.220.146:8080/tags`,tag).toPromise();
+    return tag;
+  }
+
+  async getCardsByTag(tag:Tag){
```
Comment (Lcarrico) [inline src/app/services/tag/tag.service.ts]: Please remove this function

File: src/app/services/tag/tag.service.ts [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    const cards:Card[] = await this.http.get<Card[]>(`http://34.122.220.146:8080/tags/${tag.name}/cards`).toPromise();
+    return cards;
+  }
+
+  async getStacksByTag(tag:Tag){
```
Comment (Lcarrico) [inline src/app/services/tag/tag.service.ts]: Please remove this function

**Review by Lcarrico — CHANGES_REQUESTED** [review]

**Pull request merged**